### scRNAseq analysis for Fort et al HNF4Î± Manuscript
# Related to Figures 2B-F, S2B-F, S2H, and 3B-C 

# This R script was used to identify cell types using marker genes in preprocessed scRNAseq dataset 
# This script utilizes the preprocessed seurat object generated by the 'scRNAseq_preprocessing_QC_Fort.R' script

# This code follows standard seurat workflows for Seurat V4 (https://satijalab.org/seurat/articles/pbmc3k_tutorial)


# Set working directory
setwd("/uufs/chpc.utah.edu/common/home/snydere-group2/Gabby/scRNAseq_analysis")

# Load (and install) required packages 
library(ggplot2)
#remotes::install_version("Seurat", version = "4.4.0")
library(Seurat)
#install.packages('ggpubr')
library(ggpubr)
#install.packages("R.utils")
library(R.utils)
#remotes::install_github('satijalab/seurat-wrappers@community-vignette')
library(SeuratWrappers)
#BiocManager::install("biomaRt")
library(biomaRt)


#################################################################
############# Visualize UMAP and Cell Type Markers ##############
#################################################################

# Read in seurat object generated using 'scRNAseq_preprocessing_QC_Fort.R'
KPvsKPH <- readRDS(file="all_cells_seuratobj.rds")

# Plot By Cluster
DimPlot(KPvsKPH, reduction = "umap",label=TRUE,pt.size=.3)

# Plot By Genotype
DimPlot(KPvsKPH, reduction = "umap", group.by = "Geno_Type",label=FALSE,pt.size=0.3,order=FALSE,
                     cols=c("KP"="darkcyan","KPH"="maroon"))


### Cell Type Marker DotPlot using known marker genes ###
### RELATED TO FIGURE S2C ###
levels(KPvsKPH) <- c('0','1','2','3','7','10','14','30','15','11','8','19','20',
                     '22','28','21','26','13','9','18','29','4','5','6',
                     '12','16','17','23','24','25','27')
DotPlot(KPvsKPH, features = c("Cre","Nkx2-1","Hnf4a","Foxa1","Foxa2","Foxj1","Col14a1","Col1a1","Col13a1","Pecam1",
                              "Cd3d","Cd3g","Cd3e","Cd28","Gzma","Klrc3","Cd79a","Cd79b","Fcmr","Cd19","Ly6g","S100a8","S100a9","Mmp9","Itgam","Ptprc",
                              "Itgax","Cd207","Ccl17","Ccl22")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

levels(KPvsKPH) <- c('0','1','2','3','4','5','6','7','8','9','10','11','12',
                     '13','14','15','16','17','18','19','20','21','22','23',
                     '24','25','26','27','28','29','30')


### Reassign clusters to cell names and plot
### RELATED TO FIGURE S2B ###
current_cluster_ID <- c("0","1","2","3","4","5","6","7","8","9","10","11","12",
                        "13","14","15","16","17","18","19","20","21","22","23",
                        "24","25","26","27","28","29","30")
new_cluster_ID <- c("Tumor","Tumor","Tumor","Tumor","Mac/Mono/DC","Mac/Mono/DC",
                    "Mac/Mono/DC","Tumor","T/NK","Neutrophil","Tumor","Endothelial",
                    "Mac/Mono/DC","B Cell","Tumor","Fibroblast","Mac/Mono/DC",
                    "Mac/Mono/DC","Neutrophil","T/NK","T/NK","T/NK","T/NK","Mac/Mono/DC",
                    "Mac/Mono/DC","Mac/Mono/DC","T/NK","Mac/Mono/DC","T/NK","Neutrophil",
                    "Ciliated")
KPvsKPH@active.ident <- plyr::mapvalues(x = KPvsKPH@active.ident, 
                                        from = current_cluster_ID, 
                                        to = new_cluster_ID)

cluster_plot <- DimPlot(KPvsKPH, reduction = "umap",label=TRUE,pt.size=.3)
cluster_plot + NoLegend()


###################################################################
#################### Subset out only Tumor Cells ##################
###################################################################

# Subset UMAP
KPvsKPH_tumor<-subset(KPvsKPH, idents=c("Tumor"))

# Repeat earlier steps to recluster
# Identify highly variable features
KPvsKPH_tumor <- FindVariableFeatures(KPvsKPH_tumor, selection.method = "vst", nfeatures = 2000)

#scale data
all.genes <- rownames(KPvsKPH_tumor)
KPvsKPH_tumor <- ScaleData(KPvsKPH_tumor, features = all.genes)

#Perform linear dimension reduction/PCA
KPvsKPH_tumor <- RunPCA(KPvsKPH_tumor, features = VariableFeatures(object = KPvsKPH_tumor))

# Use elbow plot to visualize dimensions that encompass most of variance
ElbowPlot(KPvsKPH_tumor, ndims=50)
# shows ~17 dims captures most of the variance

# Choosing 17 PCs for dimensionality
KPvsKPH_tumor <- FindNeighbors(KPvsKPH_tumor, dims = 1:17)
KPvsKPH_tumor <- FindClusters(KPvsKPH_tumor, resolution = 0.5)

# Run Non-Linear Dimension reduction (UMAP)
KPvsKPH_tumor <- RunUMAP(KPvsKPH_tumor, dims = 1:17)


### A small cluster of immune (ptprc+) cells were still present - cluster 8
#### Subset these out and recluster with only tumor cells
# Subset UMAP
KPvsKPH_tumor<-subset(KPvsKPH_tumor, idents=c("8"), invert = TRUE)

# Repeat earlier steps to recluster
# Identify highly variable features
KPvsKPH_tumor <- FindVariableFeatures(KPvsKPH_tumor, selection.method = "vst", nfeatures = 2000)

#Scale data
all.genes <- rownames(KPvsKPH_tumor)
KPvsKPH_tumor <- ScaleData(KPvsKPH_tumor, features = all.genes)

# Perform linear dimension reduction/PCA
KPvsKPH_tumor <- RunPCA(KPvsKPH_tumor, features = VariableFeatures(object = KPvsKPH_tumor))

# PCA plot of PC1, PC2 
DimPlot(KPvsKPH_tumor, reduction = "pca", group.by='Tumor_Type')

# Use elbow plot to visualize dimensions that encompass most of variance
ElbowPlot(KPvsKPH_tumor, ndims=50)
# shows ~12 dims captures most of the variance

# Choosing 12 PCs for dimensionality
KPvsKPH_tumor <- FindNeighbors(KPvsKPH_tumor, dims = 1:12)
KPvsKPH_tumor <- FindClusters(KPvsKPH_tumor, resolution = 0.5)

# Run Non-Linear Dimension reduction (tSNE/UMAP)
KPvsKPH_tumor <- RunUMAP(KPvsKPH_tumor, dims = 1:12)

# Save seurat object containing only tumor cells for downstream analysis
saveRDS(KPvsKPH_tumor, file="tumor_cells_seuratobj.rds")
# Run this line of code to read in this seurat object 
#KPvsKPH_tumor <- readRDS(file="tumor_cells_seuratobj.rds")





